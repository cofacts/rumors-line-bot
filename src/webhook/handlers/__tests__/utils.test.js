import {
  createPostbackAction,
  createFeedbackWords,
  createReferenceWords,
  ellipsis,
  createArticleShareBubble,
  createFlexMessageText,
  createHighlightContents,
  createReplyMessages,
} from '../utils';

describe('createArticleShareBubble()', () => {
  it('should uri size less then 1000', () => {
    const articleUrl =
      'https://cofacts.hacktabl.org/article/AWDZYXxAyCdS-nWhumlz';

    const result = createArticleShareBubble(articleUrl);
    result.footer.contents.forEach(({ action }) => {
      expect(action.uri.length).toBeLessThan(1000);
      expect(action.uri).toMatchSnapshot();
    });
  });
});

describe('ellipsis()', () => {
  it('should not ellipsis when text is short', () => {
    expect(ellipsis('12345', 10)).toBe('12345');
  });

  it('should ellipsis when text is too long', () => {
    const limit = 5;
    const processed = ellipsis('1234567890', limit);
    expect(processed).toHaveLength(limit);
    expect(processed).toMatchSnapshot();
  });

  it('should properly cut emojis', () => {
    expect(ellipsis('🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈', 5)).toMatchSnapshot();
  });

  it('should be able to customize ellipsis', () => {
    expect(ellipsis('1234567890', 5, '')).toBe('12345');
  });
});

describe('createPostbackAction()', () => {
  it('should return postback message body', () => {
    expect(
      createPostbackAction(
        '閱讀此回應',
        3,
        'I chose this',
        1519019701265,
        'some-id'
      )
    ).toMatchSnapshot();
  });
});

describe('createFeedbackWords()', () => {
  it('should create empty feedback words', () => {
    expect(createFeedbackWords(0, 0)).toMatchSnapshot();
  });
  it('should create positive feedback words', () => {
    expect(createFeedbackWords(3, 0)).toMatchSnapshot();
  });
  it('should create negative feedback words', () => {
    expect(createFeedbackWords(0, 2)).toMatchSnapshot();
  });
  it('should create both feedback words', () => {
    expect(createFeedbackWords(1, 2)).toMatchSnapshot();
  });
});

describe('createReferenceWords()', () => {
  it('should create reference for rumors', () => {
    expect(
      createReferenceWords({
        reference: 'This is a reference',
        type: 'RUMOR',
      })
    ).toMatchSnapshot();
  });
  it('should create reference for opinions', () => {
    expect(
      createReferenceWords({
        reference: 'This is refering to different opinions',
        type: 'OPINIONATED',
      })
    ).toMatchSnapshot();
  });
});

describe('createFlexMessageText', () => {
  it('should create a text for flex message', () => {
    expect(
      createFlexMessageText(
        '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我們的福利，與大家分享，可以善加利用！=7折，朋友使用ok'
      )
    ).toMatchSnapshot();
  });

  it('should handle the situation without input', () => {
    expect(createFlexMessageText()).toMatchSnapshot();
  });
});

describe('createHighlightContents', () => {
  it('should create a highlight flex message', () => {
    expect(
      createHighlightContents({ text: '<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去' })
    ).toMatchSnapshot();

    expect(
      createHighlightContents({
        text:
          '全世界有<HIGHLIGHT>成千上萬</HIGHLIGHT><HIGHLIGHT>蜜蜂</HIGHLIGHT>',
        hyperlinks: [
          {
            summary:
              '全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
            title: '<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
          },
        ],
      })
    ).toMatchSnapshot();

    expect(
      createHighlightContents({
        text:
          '全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
        hyperlinks: [
          {
            summary: null,
            title: null,
          },
        ],
      })
    ).toMatchSnapshot();

    expect(
      createHighlightContents({
        text: null,
        hyperlinks: [
          {
            summary:
              '全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
            title: '<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
          },
          {
            summary:
              '<HIGHLIGHT>計程車</HIGHLIGHT>上有裝悠遊卡<HIGHLIGHT>感應器</HIGHLIGHT>',
            title: '<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
          },
          {
            summary:
              '不影響<HIGHLIGHT>司機收入</HIGHLIGHT>，下車時使用老人悠遊卡',
            title: '下車時使用老人悠遊卡',
          },
        ],
      })
    ).toMatchSnapshot();
    expect(
      createHighlightContents({
        text: null,
        hyperlinks: [
          {
            summary: null,
            title: '<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
          },
          {
            summary: null,
            title:
              '全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>',
          },
        ],
      })
    ).toMatchSnapshot();
  });

  it('should limit letters size', () => {
    const result = createHighlightContents(
      {
        text:
          '全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
      },
      '',
      10
    );
    let lettersLength = 0;
    result.forEach(s => (lettersLength += s.text.length));
    // console.log('Letters length:' + lettersLength);
    expect(lettersLength).toBeLessThanOrEqual(10 + 3); // +3 for '...' at the end of message
  });

  it('should limit contents size', () => {
    const result = createHighlightContents(
      {
        text:
          '全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去全<HIGHLIGHT>世界</HIGHLIGHT>有成千上萬<HIGHLIGHT>蜜蜂</HIGHLIGHT>逐漸死去',
      },
      '',
      100000, // don't limit letters
      500
    );
    // console.log('Contents length:' + JSON.stringify(result).length);
    expect(JSON.stringify(result).length).toBeLessThanOrEqual(500 + 3); // +3 for '...' at the end of message
  });

  it('should handle the situation without highlight', () => {
    expect(
      createHighlightContents(undefined, 'Original text')
    ).toMatchSnapshot();
  });

  it('should handle the situation without input', () => {
    expect(createHighlightContents()).toMatchSnapshot();
  });

  it('should handle line-bot#220 issue', () => {
    expect(
      createHighlightContents(
        {
          text: null,
          hyperlinks: [],
        },
        'Original text'
      )
    ).toMatchSnapshot();
  });
});

describe('createReplyMessages()', () => {
  it('should return reply messages', () => {
    const reply = {
      type: 'RUMOR',
      reference:
        'http://www.mygopen.com/2017/06/blog-post_26.html\n神奇的地瓜葉？搭配鮮奶遠離三高？謠言讓醫生說：有痛風或是腎臟不好的人要小心！',
      text:
        '基本上地瓜葉其實單吃就有效果，牛奶、豆漿可加可不加，民眾不用迷信。 三高或是糖尿病的患者還是要搭配醫生的治療，不能單靠吃地瓜葉就想將身體調養好，民眾千萬要注意。\n另外地瓜葉內還有鉀和鈉，對於有痛風或是腎臟不好的民眾反而會造成負擔，因此並不建議食用。',
    };
    const article = { replyCount: 1 };
    const selectedArticleId = '2sn80q5l5mzi0';
    expect(
      createReplyMessages(reply, article, selectedArticleId)
    ).toMatchSnapshot();
  });
});
