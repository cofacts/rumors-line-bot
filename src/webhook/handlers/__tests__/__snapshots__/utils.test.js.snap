// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createArticleShareBubble() should uri size less then 1000 1`] = `"line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2FAWDZYXxAyCdS-nWhumlz"`;

exports[`createArticleShareBubble() should uri size less then 1000 2`] = `"https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2FAWDZYXxAyCdS-nWhumlz"`;

exports[`createFeedbackWords() should create both feedback words 1`] = `
"👍 1 user considers this helpful
😕 2 users consider this not useful"
`;

exports[`createFeedbackWords() should create empty feedback words 1`] = `"No feedback yet"`;

exports[`createFeedbackWords() should create negative feedback words 1`] = `"😕 2 users consider this not useful"`;

exports[`createFeedbackWords() should create positive feedback words 1`] = `"👍 3 users consider this helpful"`;

exports[`createFlexMessageText should create a text for flex message 1`] = `"計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我們的福利，與大家分享，可以善加利用！=7"`;

exports[`createFlexMessageText should handle the situation without input 1`] = `""`;

exports[`createHighlightContents should create a highlight flex message 1`] = `
Array [
  Object {
    "color": "#ffb600",
    "text": "蜜蜂",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "逐漸死去",
    "type": "span",
  },
]
`;

exports[`createHighlightContents should create a highlight flex message 2`] = `
Array [
  Object {
    "text": "全世界有",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "成千上萬",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "color": "#ffb600",
    "text": "蜜蜂",
    "type": "span",
    "weight": "bold",
  },
]
`;

exports[`createHighlightContents should create a highlight flex message 3`] = `
Array [
  Object {
    "text": "全",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "世界",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "有成千上萬",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "蜜蜂",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "逐漸死去",
    "type": "span",
  },
]
`;

exports[`createHighlightContents should create a highlight flex message 4`] = `
Array [
  Object {
    "text": "全",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "世界",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "有成千上萬",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "蜜蜂",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "逐漸死去
",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "計程車",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "上有裝悠遊卡",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "感應器",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "
不影響",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "司機收入",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "，下車時使用老人悠遊卡",
    "type": "span",
  },
]
`;

exports[`createHighlightContents should create a highlight flex message 5`] = `
Array [
  Object {
    "color": "#ffb600",
    "text": "蜜蜂",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "逐漸死去
全",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "世界",
    "type": "span",
    "weight": "bold",
  },
  Object {
    "text": "有成千上萬",
    "type": "span",
  },
  Object {
    "color": "#ffb600",
    "text": "蜜蜂",
    "type": "span",
    "weight": "bold",
  },
]
`;

exports[`createHighlightContents should handle line-bot#220 issue 1`] = `
Array [
  Object {
    "text": "Original text",
    "type": "span",
  },
]
`;

exports[`createHighlightContents should handle the situation without highlight 1`] = `
Array [
  Object {
    "text": "Original text",
    "type": "span",
  },
]
`;

exports[`createHighlightContents should handle the situation without input 1`] = `
Array [
  Object {
    "text": "",
    "type": "span",
  },
]
`;

exports[`createPostbackAction() should return postback message body 1`] = `
Object {
  "data": "{\\"input\\":3,\\"sessionId\\":1519019701265,\\"state\\":\\"some-id\\"}",
  "displayText": "I chose this",
  "label": "閱讀此回應",
  "type": "postback",
}
`;

exports[`createReferenceWords() should create reference for opinions 1`] = `"different opinions：This is refering to different opinions"`;

exports[`createReferenceWords() should create reference for rumors 1`] = `"references：This is a reference"`;

exports[`createReplyMessages() should return reply messages 1`] = `
Array [
  Object {
    "text": "💡 Someone on the internet replies to the message:",
    "type": "text",
  },
  Object {
    "text": "基本上地瓜葉其實單吃就有效果，牛奶、豆漿可加可不加，民眾不用迷信。 三高或是糖尿病的患者還是要搭配醫生的治療，不能單靠吃地瓜葉就想將身體調養好，民眾千萬要注意。
另外地瓜葉內還有鉀和鈉，對於有痛風或是腎臟不好的民眾反而會造成負擔，因此並不建議食用。",
    "type": "text",
  },
  Object {
    "text": "references：http://www.mygopen.com/2017/06/blog-post_26.html
神奇的地瓜葉？搭配鮮奶遠離三高？謠言讓醫生說：有痛風或是腎臟不好的人要小心！",
    "type": "text",
  },
  Object {
    "text": "⬆️ Therefore, the author think the message contains misinformation.

💁 This content is provided by Cofact message reporting chatbot and crowd-sourced fact-checking community under CC BY-SA 4.0 license. Please refer to their references and make judgements on your own.


⁉️ If you have different thoughts, you may have your say here:
https://dev.cofacts.tw/article/2sn80q5l5mzi0",
    "type": "text",
  },
]
`;

exports[`createTypeWords should return the type words for NOT_ARTICLE 1`] = `"Invalid request"`;

exports[`createTypeWords should return the type words for NOT_RUMOR 1`] = `"Contains true information"`;

exports[`createTypeWords should return the type words for OPINIONATED 1`] = `"Contains personal perspective"`;

exports[`createTypeWords should return the type words for RUMOR 1`] = `"Contains misinformation"`;

exports[`createTypeWords should return the type words for other types 1`] = `"Undefined"`;

exports[`ellipsis() should ellipsis when text is too long 1`] = `"123⋯⋯"`;

exports[`ellipsis() should properly cut emojis 1`] = `"🏳️‍🌈🏳️‍🌈🏳️‍🌈⋯⋯"`;
